[Console] Add hidden and deprecated option modes

flkasper opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 8.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | Fix #54206
| License       | MIT

Adds two new modes to `InputOption`: `InputOption::HIDDEN` and `InputOption::DEPRECATED`. They combine with the existing `VALUE_*` modes.

```php
$this
    ->addOption('old', 'o', InputOption::VALUE_REQUIRED | InputOption::DEPRECATED, 'Use --new instead')
    ->addOption('internal', null, InputOption::VALUE_NONE | InputOption::HIDDEN, 'Internal switch')
;
```

The same is available on the `#[Option]` attribute of invokable commands:

```php
public function __invoke(
    #[Option(deprecated: true)] ?string $old = null,
    #[Option(hidden: true)] bool $internal = false,
): int {
}
```

## Hidden options

A hidden option works exactly like any other option. It is only left out of the places that advertise the option to the user:

 * the command synopsis, so `[--internal]` is not part of `demo [options]`;
 * every descriptor, so `help`, `list` and their `--format=json|xml|md|rst` variants skip it;
 * shell completion, both as a suggestion and as a known option when resolving what the cursor sits on.

Use it for switches an application passes to itself, or for options kept for compatibility that should no longer be advertised.

`help <command> --show-hidden-options` lists them back. In the text output they carry a `[hidden]` prefix, so a hidden option is still recognizable when shown:

```
Options:
  -o, --old=OLD         [deprecated] Use --new instead
      --new=NEW         The new option
      --internal        [hidden] Internal switch
```

## Deprecated options

A deprecated option stays fully usable. It is marked in the help output and it warns when it is actually used:

 * the text descriptor greys out the synopsis and prefixes the description with `[deprecated]`;
 * passing the option, by name or by shortcut, writes a warning block on the error output before the command runs.

```
$ demo --old=1 > out.txt

  The option "--old|-o" is deprecated.

```

The warning goes to stderr, so it never pollutes the command output. It is only written when the option is really passed on the command line, not when it falls back to its default value.

## Descriptor output

The `json`, `xml`, `md` and `rst` descriptors gained two entries per option, always present:

```json
{
    "name": "--old",
    "shortcut": "-o",
    "accept_value": true,
    "is_value_required": true,
    "is_multiple": false,
    "is_deprecated": true,
    "is_hidden": false,
    "description": "Use --new instead",
    "default": null
}
```

## Points for the documentation

 * `InputOption::HIDDEN` and `InputOption::DEPRECATED` are modes, so they are combined with `VALUE_NONE`, `VALUE_REQUIRED`, `VALUE_OPTIONAL`, `VALUE_IS_ARRAY` and `VALUE_NEGATABLE` with `|`, exactly like the existing ones. Used alone they imply `VALUE_NONE`.
 * `#[Option(deprecated: true)]` and `#[Option(hidden: true)]` are the equivalents for invokable commands.
 * `InputOption::isHidden()` and `InputOption::isDeprecated()` report the two flags.
 * Hidden is about visibility only. It never blocks the option, and it is not a security measure.
 * `help <command> --show-hidden-options` is the way to see hidden options. `--show-hidden-options` is itself hidden.
 * Deprecating an option is a two-step story: mark it `DEPRECATED` in one minor version, remove it in the next major one.
 * The deprecation warning is written on stderr and is not a PHP deprecation, so it is not captured by an error handler nor reported by the deprecation collector.

flkasper committed:
[Console] Add of hidden and deprecation option flags

flkasper committed:
Fix code style

94noni commented:
Waw very nice :)
Passing by, i would just split the PR commits into 2 separated PRs
Feature addition and cs fix on the other

flkasper commented:
@94noni I actually only wanted to fix my commit via ecs and had therefore limited ecs to the console bundle only.
Who would have thought that the code would differ in so many places :)

94noni commented:
Got it :)
Lets wait other reviews 👌🏼

smnandre commented:
Love the idea... but i feel really agressed by the red, and i feel it makes the option an unwanted point of attention.

Did you try with other colors ? Something muted maybe to get the opposite effect ? 

I think we should not rely _only_ on colors (for accessibility reasons but also technical ones), 

What about an automatic `[deprecated]` prefix in the description ? 

```
Options:
    -test-deprecated            [deprecated] Lorem ipsum ...

```    

With `[deprecated]` not having to be set in the option description

flkasper commented:
> Love the idea... but i feel really agressed by the red, and i feel it makes the option an unwanted point of attention.
> Did you try with other colors ? Something muted maybe to get the opposite effect ?

I had originally thought of yellow (warning), but this is already used in the headings such as "Options:".

I have now simply output in every available style and colour for comparison.
Personally, I think that yellow, grey and white would suit both in light and dark mode.

![image](https://github.com/symfony/symfony/assets/26827635/a8f9b33a-88b8-421b-a48c-647b986f49f5)
![image](https://github.com/symfony/symfony/assets/26827635/9384a9fd-b5c5-419b-b08a-fb0f46a7b987)

flkasper commented:
> What about an automatic [deprecated] prefix in the description ?

I had also thought of this and it is actually a nice idea, but it has the disadvantage that you cannot simply customise this prefix.
See screenshots for the colours.

You would then have to add something like this (either to set the prefix or the message):

```PHP
private array $optionDeprecationPrefix = [];

public function setOptionDeprecationPrefix(string $optionName, string $deprecationPrefix) {
    $this->optionDeprecationPrefix[$optionName] = $deprecationPrefix;
}

// On output:
$deprecationPrefix =  $this->optionDeprecationPrefix[$optionName] ??  'deprecated';
" [$deprecationPrefix] $optionDescription"
```

smnandre commented:
> [...] grey [...] 

Would be my personal choices, but colors are subjective, so let's see what others think about it ? :) 

> the disadvantage that you cannot simply customise this prefix.

That was not (in my mind) something configurable so i did not add the constraint. 

But if you have commands from your app and bundles, that would be better to not allow a different style per command, no ? 


I guess i'd do something like 

```console
sf list 

lorem:list            [deprecated] List all lorem from ipsum


sf list -v

lorem:list            [deprecated since 2.0] List all lorem from ipsum


sf list -vv

lorem:list            [deprecated since 2.0] List all lorem from ipsum  [Use foo:bar instead]
```

flkasper committed:
Change style of deprecated option to inline formatting

GromNaN commented:
I like the grey too. We want this options to be less visible in the list, not highlight them.

And I don't see any reason for customizing the description prefix either. Consistency across commands is better.

flkasper committed:
[Console] Change color of deprecated option to gray

flkasper committed:
[Console] Fixed test after color change

GromNaN commented on src/Symfony/Component/Console/Command/Command.php:
```diff
@@ -648,6 +654,29 @@ public function getHelper(string $name): HelperInterface
+                try {
+                    $optionName = $inputOption->getName();
+                    $optionValue = $input->getOption($optionName);
+                    if (isset($optionValue) && $optionValue !== $inputOption->getDefault()) {
```
The deprecation message is not displayed when the option is set to it's default value.
It would be nice to be able to detect whether an option is set to always display the depreciation message.

As this, people that use the deprecated option with the default value will have breaking change without notice when the deprecated option is removed.

flkasper replied:
> It may take a deeper work to detect if an option has been passed. This could be a feature in its own.

As you have already written, it is currently not possible to recognise whether an option has been used or whether the default value is simply being used.

Hence the check against the default value

GromNaN replied:
You can use `$input->hasParameterOption()` to check if an option is provided:

```php
$names = ['--'.$inputOption->getName()];
if ($inputOption->getShortcut()) {
    $names[] = '-'.$inputOption->getShortcut();
}
if ($input->hasParameterOption($names, true)) {
    $deprecationMessages[] = sprintf('The option "%s" is deprecated.', implode('|', $names));
}
```

flkasper replied:
@GromNaN's suggested change applied

GromNaN commented on src/Symfony/Component/Console/Completion/CompletionInput.php:
```diff
@@ -226,7 +234,7 @@ private function isCursorFree(): bool
     }
 
-    public function __toString()
+    public function __toString(): string
```
Good catch. This is fine since the class is final.

GromNaN commented on src/Symfony/Component/Console/Descriptor/Descriptor.php:
```diff
@@ -42,6 +42,26 @@ public function describe(OutputInterface $output, object $object, array $options
+     */
+    protected function skipHiddenOption(InputOption $inputOption, array $options = []): bool
+    {
+        return $inputOption->isHidden() && !($options['show-hidden-options'] ?? false);
```
This method can be removed if you check `$options['show-hidden-options']` before the `array_filter`. The code will be simpler.

chalasr replied:
The method has more usages though

GromNaN replied:
Indeed.

GromNaN commented on src/Symfony/Component/Console/.editorconfig:
This file is unrelated. If you have a strong feeling that it should be added, it must be in a distinct PR with a description why they are useful.

flkasper replied:
File removed.
Was intended to ensure that when saving the fixtures the trailing whitespace is not trimmed.

Let's see if I do a PR for it.

GromNaN commented on src/Symfony/Component/Console/Command/Command.php:
```diff
@@ -648,6 +654,29 @@ public function getHelper(string $name): HelperInterface
+                    $optionName = $inputOption->getName();
+                    $optionValue = $input->getOption($optionName);
+                    if (isset($optionValue) && $optionValue !== $inputOption->getDefault()) {
+                        $deprecationMessages[] = sprintf('The option "--%s" is deprecated.', $optionName);
```
Non-blocker: if people use the shortcut, this deprecation message mention the full name of the option. They have to look at the command definition to find which option have been used.

flkasper replied:
I have included the shortcut in the output, if it exists.

GromNaN commented on src/Symfony/Component/Console/Helper/ProgressIndicator.php:
```diff
@@ -54,7 +54,6 @@ public function __construct(
         private int $indicatorChangeInterval = 100,
         ?array $indicatorValues = null,
     ) {
-
```
Please revert this unrelated change.

GromNaN commented on src/Symfony/Component/Console/Input/InputOption.php:
```diff
@@ -50,6 +50,22 @@ class InputOption
     public const VALUE_NEGATABLE = 16;
 
+    /**
+     * The option allows to deprecate command option.
+     * Option will be marked as deprecated in help output and a warning is printed when the command is executed.
```
```suggestion
     * Mark the option as deprecated in help output. A message is printed when the command is executed.
```

GromNaN commented on src/Symfony/Component/Console/Tests/Fixtures/Style/SymfonyStyle/command/command_4.php:
Please revert unrelated changes in the fixtures files.

GromNaN commented on src/Symfony/Component/Console/Tests/Fixtures/input_option_hidden.json:
```diff
@@ -0,0 +1 @@
+NULL
```
Why is there a `NULL` char here?

flkasper replied:
I have extended the existing test providers with corresponding test cases for hidden options.

NULL is a valid json (string) and unfortunately necessary, as otherwise JsonDescriptorTest::testDescribeInputOption would fail.

GromNaN commented on src/Symfony/Component/Console/Tests/Helper/TableTest.php:
```diff
@@ -102,7 +102,7 @@ public static function renderProvider()
                 $books,
                 'default',
-<<<'TABLE'
+                <<<'TABLE'
```
Unrelated change to be reverted too.

GromNaN reviewed (commented):
I really like this feature. I did a few tests on an application. It works very well, congratulations for your work.

There's a comment about the depreciation message display. It may take a deeper work to detect if an option has been passed. This could be a feature in its own.

The large number of style changes unrelated to the PR make review more difficult. These changes should be made in dedicated PRs with arguments. They can have an impact on maintenance (and upmerges from old branches to new ones).

flkasper committed:
[Console] Revert unrelated (style) changes

flkasper committed:
[Console] Apply suggestion changes

- Included option shortcut in deprecation message
- Applied phpdoc suggestion
- Applied remove of console .editorconfig

smnandre commented on src/Symfony/Component/Console/Input/InputOption.php:
```diff
@@ -259,4 +294,14 @@ public function equals(self $option): bool
+     *
+     * @return bool true if mode is $mode, false otherwise
+     */
+    protected function hastMode(int $mode): bool
```
Is this different from 'hasMode' ? I may not know the term

flkasper replied:
I just missed the t in the change from `getMode` to `hasMode`

smnandre replied:
I really did not know 😅

flkasper committed:
[Console] Fixed name of method

GromNaN commented on src/Symfony/Component/Console/Tests/Fixtures/Foo1Command.php:
```diff
@@ -1,5 +1,14 @@
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
```
Is it php-cs-fixer that caused this license block to be added? That can't be done in this PR because it makes it look like there are a lot more modified files than actually necessary for this feature. 
Please revert all formatting changes not related to the PR.

stof replied:
the config of php-cs-fixer should be ignoring all files in `Fixtures` folder, so I doubt it is the case (unless a custom config was used, which would be a no-go)

flkasper replied:
> Is it php-cs-fixer that caused this license block to be added?

Yes, this is due to the php-cs-fixer.

I have used the config .php-cs-fixer.dist.php from the root.
Rule [header_comment](https://github.com/symfony/symfony/blob/9b323c6866078f1bc761d7cc34297c41815108d3/.php-cs-fixer.dist.php#L32) is being configured there.
No idea why the fixtures are also changed.

```
#Executed command:
php tools/php-cs-fixer/vendor/bin/php-cs-fixer fix --config=.php-cs-fixer.dist.php <folder>
```

I'll look over it again and remove all unrelated changes to the PR.

smnandre replied:
When you pass a directory, you need to use `--path-mode=intersection` (from memory, maybe some typo there) or phpcs ignores the previously defined paths .. and exclusions.

smnandre replied:
Indeed sorry * 

```console
php php-cs-fixer.phar fix --path-mode=intersection /path/to/dir
```
https://cs.symfony.com/doc/usage.html#the-fix-command

(* my local alias for php-cs-fixer is... `phpcs` 😶 )

stof commented:
I haven't checked the current state of the PR yet. But for sure, we cannot rely on color only:

- color blind people won't see them, causing accessibility issues
- not all terminals support colors

flkasper committed:
[Console] Revert unrelated (style) changes

flkasper committed:
[Console] Applied change suggestion in Command::writeDeprecationMessages

flkasper committed:
[Console] Added suggested description prefix to deprecated options in TextDescriptor.php

flkasper commented:
I have added @smnandre's suggestion for "[deprecated] prefix to the option description in the TextDescriptor.

GromNaN commented on src/Symfony/Component/Console/Input/InputOption.php:
```diff
@@ -50,6 +50,21 @@ class InputOption
+    public const DEPRECATED = 32;
+
+    /**
+     * The option allows to hide command option from help.
```
```suggestion
     * Hide the option from command descriptors.
```

flkasper replied:
Applied

GromNaN reviewed (approved):
LGTM, with a last suggestion.

flkasper committed:
[Console] Apply suggestion & fixed coding standard of changes

nicolas-grekas commented on src/Symfony/Component/Console/Command/Command.php:
```diff
@@ -273,6 +275,10 @@ public function run(InputInterface $input, OutputInterface $output): int
 
         $input->validate();
 
+        if (isset($inputDefinition)) {
```
why the isset?
```suggestion
        if ($inputDefinition) {
```

GromNaN replied:
`$inputDefinition` is defined in a try..catch, but should be outside.

flkasper replied:
As @GromNaN has already explained, `$inputDefinition` is defined within a try...catch:
```
try {
    $inputDefinition = $this->getDefinition();
    $input->bind($inputDefinition);
} catch (ExceptionInterface $e) {
    if (!$this->ignoreValidationErrors) {
        throw $e;
    }
}
```
Since `getDefinition` can also throw an exception (if the parent constructor call is missing), moving it outside the try...catch makes little sense.

Therefore, `isset` must be used to check whether `$inputDefinition` is defined.

nicolas-grekas commented on src/Symfony/Component/Console/Command/Command.php:
```diff
@@ -648,6 +654,27 @@ public function getHelper(string $name): HelperInterface
+                }
+            }
+        }
+        if (!empty($deprecationMessages)) {
```
```suggestion
        if ($deprecationMessages) {
```

nicolas-grekas commented on src/Symfony/Component/Console/Command/Command.php:
```diff
@@ -648,6 +654,27 @@ public function getHelper(string $name): HelperInterface
+                    $optionNames[] = '-'.$inputOption->getShortcut();
+                }
+                if ($input->hasParameterOption($optionNames, true)) {
+                    $deprecationMessages[] = sprintf('The option "%s" is deprecated.', implode('|', $optionNames));
```
```suggestion
                    $deprecationMessages[] = \sprintf('The option "%s" is deprecated.', implode('|', $optionNames));
```

nicolas-grekas commented on src/Symfony/Component/Console/Descriptor/TextDescriptor.php:
```diff
@@ -73,41 +73,47 @@ protected function describeInputOption(InputOption $option, array $options = [])
         $spacingWidth = $totalWidth - Helper::width($synopsis);
 
-        $this->writeText(sprintf('  <info>%s</info>  %s%s%s%s',
+        $synopsis = sprintf('<%1$s>%2$s</%1$s>', $option->isDeprecated() ? 'fg=gray;' : 'info', $synopsis);
```
```suggestion
        $synopsis = \sprintf('<%1$s>%2$s</%1$s>', $option->isDeprecated() ? 'fg=gray;' : 'info', $synopsis);
```

nicolas-grekas commented on src/Symfony/Component/Console/Descriptor/TextDescriptor.php:
```diff
@@ -73,41 +73,47 @@ protected function describeInputOption(InputOption $option, array $options = [])
+
+        $description = trim(($option->isDeprecated() ? '[deprecated] ' : '').$option->getDescription());
+
+        $this->writeText(sprintf('  %s  %s%s%s%s',
```
```suggestion
        $this->writeText(\sprintf('  %s  %s%s%s%s',
```

nicolas-grekas commented on src/Symfony/Component/Console/Tests/Input/InputOptionTest.php:
```diff
@@ -83,34 +83,67 @@ public function testModes()
+    public static function privideInvalidModeData()
+    {
+        yield 'negative mode' => [-1, 'Option mode "-1" is not valid.'];
+        yield 'invalid bit mask value' => [InputOption::HIDDEN << 1, sprintf('Option mode "%d" is not valid.', InputOption::HIDDEN << 1)];
```
```suggestion
        yield 'invalid bit mask value' => [InputOption::HIDDEN << 1, \sprintf('Option mode "%d" is not valid.', InputOption::HIDDEN << 1)];
```

nicolas-grekas commented on src/Symfony/Component/Console/Descriptor/MarkdownDescriptor.php:
```diff
@@ -81,19 +82,20 @@ protected function describeInputDefinition(InputDefinition $definition, array $o
         }
 
-        if (\count($definition->getOptions()) > 0) {
+        $inputOptions = $this->removeHiddenOptions($definition->getOptions(), $options);
+        if (!empty($inputOptions)) {
```
```suggestion
        if ($inputOptions = $this->removeHiddenOptions($definition->getOptions(), $options)) {
```

nicolas-grekas commented on src/Symfony/Component/Console/Descriptor/TextDescriptor.php:
```diff
@@ -73,41 +73,47 @@ protected function describeInputOption(InputOption $option, array $options = [])
         }
 
-        if ($definition->getArguments()) {
+        $showArguments = !empty($inputArguments);
+        if ($showArguments) {
```
```suggestion
        if ($inputArguments) {
```

nicolas-grekas commented on src/Symfony/Component/Console/Descriptor/TextDescriptor.php:
```diff
@@ -73,41 +73,47 @@ protected function describeInputOption(InputOption $option, array $options = [])
-        }
-
-        if ($definition->getOptions()) {
+        if (!empty($inputOptions)) {
+            if ($showArguments) {
```
```suggestion
        if ($inputOptions) {
            if ($inputArguments) {
```

nicolas-grekas commented on src/Symfony/Component/Console/Input/InputOption.php:
```diff
@@ -259,4 +294,14 @@ public function equals(self $option): bool
+     *
+     * @return bool true if mode is $mode, false otherwise
+     */
+    protected function hasMode(int $mode): bool
```
I'd rather not add a new method, it's unrelated to the proposed feature

flkasper replied:
Is an internal function and simplifies and reduces duplicate/similar code

stof replied:
This method is not added as an internal method. You added it as part of the semver API.

If you want to keep the helper, make it private instead.

nicolas-grekas reviewed (changes requested):
LGTM, I just have minor comments. Can you please rebase also?

flkasper committed:
Apply suggestions from code review

Co-authored-by: Nicolas Grekas <nicolas.grekas@gmail.com>

Luc45 commented on src/Symfony/Component/Console/Tests/Fixtures/application_2.json:
```diff
@@ -765,6 +765,93 @@
+                        "description": "Do not ask any interactive question",
+                        "default": false
+                    },
+                    "deprecated_option": {
```
`deprecated_option` should probably be `deprecated-option`?

Luc45 replied:
To follow the pattern of the other options in this snapshot test: `no-interaction` and `no-ansi`

flkasper replied:
> `deprecated_option` should probably be `deprecated-option`?

I have used an option with name in snake_case format in the TestCommand, so it is correct.
SnakeCase is easier for copy/paste.

flkasper committed:
Cleanup after rebase

flkasper committed:
Apply cs patch

stof commented on src/Symfony/Component/Console/Descriptor/MarkdownDescriptor.php:
```diff
@@ -70,6 +70,7 @@ protected function describeInputOption(InputOption $option, array $options = [])
             .'* Accept value: '.($option->acceptValue() ? 'yes' : 'no')."\n"
             .'* Is value required: '.($option->isValueRequired() ? 'yes' : 'no')."\n"
             .'* Is multiple: '.($option->isArray() ? 'yes' : 'no')."\n"
+            .($option->isDeprecated() ? ('* Is deprecated: yes'."\n") : '')
```
shouldn't we have a line `Is deprecated: no` instead ?

stof commented on src/Symfony/Component/Console/Descriptor/JsonDescriptor.php:
```diff
@@ -115,20 +119,26 @@ private function getInputOptionData(InputOption $option, bool $negated = false):
             'default' => \INF === $option->getDefault() ? 'INF' : $option->getDefault(),
         ];
+        if (!$option->isDeprecated()) {
+            unset($data['is_deprecated']);
```
I would not unset this. The json is easier to use if we don't have conditional keys in it.

flkasper committed:
[Console] Add hidden and deprecated option flags

Add the `InputOption::HIDDEN` and `InputOption::DEPRECATED` modes, along
with `InputOption::isHidden()` and `InputOption::isDeprecated()`, and the
matching `hidden` and `deprecated` arguments of the `#[Option]` attribute.

Hidden options are left out of the command synopsis, of the descriptors and
of shell completion. `help --show-hidden-options` lists them back.

Deprecated options are greyed out and prefixed with `[deprecated]` in the
text descriptor, and using one writes a warning block on the error output.

The json, xml, markdown and reStructuredText descriptors gained the
`is_deprecated` and `is_hidden` entries.
